vyos/vyos-1x
# T6948: Keep DHCP server leases in sync with hostd records

## Change Summary

Keep DHCP server leases in sync with vyos-hostd records via helper script invoked with `ExecStartPost` directive in kea-dhcp4-server.service.
 
The helper script updates VyOS hostd records from DHCP server leases. This ensures that hostd records with the DHCP server leases are kept in sync with VyOS hostd records right after DHCP server is started.

This is similar to the capability exposed via kea hook `libdhcp_run_script.so` which is invoked internally by kea when a single lease changes state.
 
Since the kea hook is currently implemented for DHCPv4 only, this helper script is implemented for DHCPv4 only as well.

## Types of changes


## Related Task(s)

* https://vyos.dev/T6948

## Related PR(s)

https://github.com/vyos/vyos-1x/pull/4307

## Component(s) name

dhcp server

## Proposed changes


## How to test

1. Enable DHCP server and have active server leases
2. Ensure that relevant host entries exist in `/etc/hosts` and/or `/run/vyos-hostsd/vyos-hostsd.state`
4. Restart VyOS
5. Verify DHCP server leases are still active (`show dhcp server leases`)
6. Verify that the leases have NOT been applied to `/etc/hosts`
7. Restart dhcp server (`restart dhcp server`)
8. Verify that the leases have been applied to `/etc/hosts`

## Smoketest result

```
vyos@test-1216:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_dhcp-server.py
test_dhcp_exclude_in_range (__main__.TestServiceDHCPServer.test_dhcp_exclude_in_range) ... ok
test_dhcp_exclude_not_in_range (__main__.TestServiceDHCPServer.test_dhcp_exclude_not_in_range) ... ok
test_dhcp_high_availability (__main__.TestServiceDHCPServer.test_dhcp_high_availability) ... ok
test_dhcp_high_availability_standby (__main__.TestServiceDHCPServer.test_dhcp_high_availability_standby) ... ok
test_dhcp_hostsd_lease_sync (__main__.TestServiceDHCPServer.test_dhcp_hostsd_lease_sync) ... ok
test_dhcp_multiple_pools (__main__.TestServiceDHCPServer.test_dhcp_multiple_pools) ... ok
test_dhcp_on_interface_with_vrf (__main__.TestServiceDHCPServer.test_dhcp_on_interface_with_vrf) ... ok
test_dhcp_relay_server (__main__.TestServiceDHCPServer.test_dhcp_relay_server) ... ok
test_dhcp_single_pool_options (__main__.TestServiceDHCPServer.test_dhcp_single_pool_options) ... ok
test_dhcp_single_pool_options_scoped (__main__.TestServiceDHCPServer.test_dhcp_single_pool_options_scoped) ... ok
test_dhcp_single_pool_range (__main__.TestServiceDHCPServer.test_dhcp_single_pool_range) ... ok
test_dhcp_single_pool_static_mapping (__main__.TestServiceDHCPServer.test_dhcp_single_pool_static_mapping) ... ok

```

## Checklist:
